compiler/frontend/src/hir_to_mir.rs

```diff
@@ -549,55 +521,111 @@ impl<'a> LoweringContext<'a> {
                 // TODO: concat reasons
                 body.push_panic(reason, responsible_for_match)
             }
-            [(case_pattern, case_body), rest @ ..] => {
+            [hir::MatchCase {
+                pattern: case_pattern,
+                identifier_expressions: case_identifiers,
+                condition: case_condition,
+                body: case_body,
+            }, rest @ ..] => {
                 let pattern_result = PatternLoweringContext::compile_pattern(
                     body,
                     hir_id.clone(),
                     responsible_for_match,
                     expression,
                     case_pattern,
                 );
+                let builtin_if_else = body.push_builtin(BuiltinFunction::IfElse);
 
-                let is_match = body.push_is_match(pattern_result, responsible_for_match);
-
+                let is_pattern_match = body.push_is_match(pattern_result, responsible_for_match);
                 let case_id = hir_id.child(format!("case-{case_index}"));
-                let builtin_if_else = body.push_builtin(BuiltinFunction::IfElse);
-                let then_function = body.push_function(case_id.child("matched"), |body, _| {
-                    self.ongoing_destructuring = Some(OngoingDestructuring {
-                        result: pattern_result,
-                        is_trivial: false,
-                    });
-                    self.compile_expressions(body, responsible_for_needs, &case_body.expressions);
-                });
-                let else_function = body.push_function(case_id.child("didNotMatch"), |body, _| {
-                    let list_get_function = body.push_builtin(BuiltinFunction::ListGet);
-                    let one = body.push_int(1);
-                    let reason = body.push_call(
-                        list_get_function,
-                        vec![pattern_result, one],
-                        responsible_for_match,
-                    );
-                    no_match_reasons.push(reason);
 
+                let else_function = body.push_function(case_id.child("didNotMatch"), |body, _| {
```

Comment:
Note: The else function is now pushed before the then function, which makes reading the MIR a little more convoluted.